src/txn/txn.c

```diff
@@ -543,22 +544,39 @@ __wt_txn_config(WT_SESSION_IMPL *session, const char *cfg[])
      * Check if the prepare timestamp and the commit timestamp of a prepared transaction need to be
      * rounded up.
      */
-    WT_RET(__wt_config_gets_def(session, cfg, "roundup_timestamps.prepared", 0, &cval));
+    WT_ERR(__wt_config_gets_def(session, cfg, "roundup_timestamps.prepared", 0, &cval));
     if (cval.val)
         F_SET(txn, WT_TXN_TS_ROUND_PREPARED);
 
     /* Check if read timestamp needs to be rounded up. */
-    WT_RET(__wt_config_gets_def(session, cfg, "roundup_timestamps.read", 0, &cval));
+    WT_ERR(__wt_config_gets_def(session, cfg, "roundup_timestamps.read", 0, &cval));
     if (cval.val)
         F_SET(txn, WT_TXN_TS_ROUND_READ);
 
-    WT_RET(__wt_config_gets_def(session, cfg, "read_timestamp", 0, &cval));
+    /* Check if the read timestamp is allowed to be before the oldest timestamp. */
+    WT_ERR(__wt_config_gets_def(session, cfg, "read_before_oldest", 0, &cval));
+    if (cval.val) {
+        if (F_ISSET(txn, WT_TXN_TS_ROUND_READ))
+            WT_ERR_MSG(session, EINVAL,
+              "cannot specify roundup_timestamps.read and "
+              "read_before_oldest on the same transaction");
+        F_SET(txn, WT_TXN_TS_READ_BEFORE_OLDEST);
+    }
+
+    WT_ERR(__wt_config_gets_def(session, cfg, "read_timestamp", 0, &cval));
     if (cval.len != 0) {
-        WT_RET(__wt_txn_parse_timestamp(session, "read", &read_ts, &cval));
-        WT_RET(__wt_txn_set_read_timestamp(session, read_ts));
+        WT_ERR(__wt_txn_parse_timestamp(session, "read", &read_ts, &cval));
+        WT_ERR(__wt_txn_set_read_timestamp(session, read_ts));
     }
 
-    return (0);
+err:
+    if (ret != 0)
+        /*
+         * In the event that we error during configuration we should clear the flags on the
+         * transaction so they are not set in a subsequent call to transaction begin.
+         */
+        txn->flags = 0;
```

Comment:
What about `txn->isolation` and `txn->txn_logsync`? These also get set in this function and won't get unset in an error.

Maybe it's better to do most of the `__wt_txn_release` logic in the error case.

Reply:
This should be sufficient because `isolation` and `txn_logsync` are initialized at the start of `wt_txn_begin` before any call to this config function with the values they inherit from the session and connection respectively.

Reply:
Makes sense. Thanks. 👍